### ISSUE TITLE
Kubernetes - add SAN for planetary network IP

### ISSUE THREAD

**Issue by scottyeager** [threefoldtech/tf-images#98]
Under the current configuration, nodes' planetary network IPs don't get added to the certificate that's generated to authenticate `kubectl` interactions via TLS. Therefore, the user will see an error like this if they attempt to control the cluster over planetary network:

```
Unable to connect to the server: x509: certificate is valid for OVERLAY_IP, 127.0.0.1 not PLANETARY_IP
```

Additional addresses can be added to the certificate using the `--tls-san` option when launching `k3s`, as documented [here](https://docs.k3s.io/reference/server-config#listeners).

I didn't test with a public IP but I guess it will have the same issue. Each additional IP should be added to allow access on any interface.

**Comment by jakubprogramming** [comment]
Temporary workaround that I used to update the certificate (on host machine of cluster):
`$ k3s server --tls-san [PLANETARY_IP]:6443`

**Comment by sameh-farouk** [comment]
I was checking this, and k3s By default seems to add a bunch of SANs.
trying to reproduce this issue, I did a copy of the k3s.yaml to my local machine, edited the server IP, and replaced it with the planetary IP. I was able to connect without issues.

```bash
❯ kubectl get all
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.43.0.1    <none>        443/TCP   7h11m
```

To verify what SANs were added on a k3s server you can do `kubectl -n kube-system get secret k3s-serving -o yaml`.
I checked this. here is the output

```bash
❯ kubectl -n kube-system get secret k3s-serving -o yaml
apiVersion: v1
data:
...
kind: Secret
metadata:
  annotations:
    ...
    listener.cattle.io/cn-10.20.4.2: 10.20.4.2
    listener.cattle.io/cn-10.43.0.1: 10.43.0.1
    listener.cattle.io/cn-127.0.0.1: 127.0.0.1
    listener.cattle.io/cn-301_1037_16ee_23fb_85c3_6980_1999_-8763dc: 301:1037:16ee:23fb:85c3:6980:1999:74c6
    listener.cattle.io/cn-kubernetes: kubernetes
    listener.cattle.io/cn-kubernetes.default: kubernetes.default
    listener.cattle.io/cn-kubernetes.default.svc: kubernetes.default.svc
    listener.cattle.io/cn-kubernetes.default.svc.cluster.local: kubernetes.default.svc.cluster.local
    listener.cattle.io/cn-localhost: localhost
    listener.cattle.io/cn-mrfdb09c16: mrfdb09c16
    listener.cattle.io/fingerprint: SHA1=39A9C17000914271D6CB2C1DF14E7F33B60F97E0
  creationTimestamp: "2023-01-18T16:00:58Z"
  name: k3s-serving
  namespace: kube-system
  resourceVersion: "6603"
  uid: 2bdb5edf-b8e5-43e0-a41e-54ff15e249c0
type: kubernetes.io/tls
```

As you see the planetary network IP was added to the annotations list of secret k3s-serving

I'll try to reproduce it one more time on a fresh deployment and report back. maybe I changed something previously.

**Comment by sameh-farouk** [comment]
I can reporoduce it in a fresh deployment. this is wired not sure how is was added in the previous deployment.
```bash
root@MR4a395e97:~# kubectl -n kube-system get secret k3s-serving -o yaml
apiVersion: v1
data:
  tls.crt: ...
  tls.key: ...
kind: Secret
metadata:
  annotations:
    listener.cattle.io/cn-10.20.4.2: 10.20.4.2
    listener.cattle.io/cn-10.43.0.1: 10.43.0.1
    listener.cattle.io/cn-127.0.0.1: 127.0.0.1
    listener.cattle.io/cn-kubernetes: kubernetes
    listener.cattle.io/cn-kubernetes.default: kubernetes.default
    listener.cattle.io/cn-kubernetes.default.svc: kubernetes.default.svc
    listener.cattle.io/cn-kubernetes.default.svc.cluster.local: kubernetes.default.svc.cluster.local
    listener.cattle.io/cn-localhost: localhost
    listener.cattle.io/cn-mr4a395e97: mr4a395e97
    listener.cattle.io/fingerprint: SHA1=24128BD6FDEE7B54635A02B3A11A69B48B1B97E4
  creationTimestamp: "2023-01-19T14:19:57Z"
  name: k3s-serving
  namespace: kube-system
  resourceVers
```

```bash
❯ kubectl get all
Unable to connect to the server: x509: certificate is valid for 10.20.4.2, 10.43.0.1, 127.0.0.1, not 301:1037:16ee:23fb:7c5b:9fc7:be13:6eff
```

**Comment by sameh-farouk** [comment]
The cert is a bit odd in that it is dynamically updated on-demand to add valid addresses so it's not surprising that you might not see what you want right away if you're not actually hitting it with traffic at that address yet.

You should see additional DNS or IP SAN entries show up on the port 6443 listener after you have made a request to that IP or hostname.

The reason i had the plantery ip in the listner is that i tried in the previous deployment to curl `/ping` end point from my local machine.
  
```bash
curl -v https://[<plantery ip >]:6443/ping
```

then kubectl stoped complaining.

**Comment by sameh-farouk** [comment]
beside knowing this, I will upadete it from the server side in the entrypoint, for added convince.

**Comment by sameh-farouk** [comment]
Update:
PR ready for review
https://github.com/threefoldtech/tf-images/pull/122

**Comment by sameh-farouk** [comment]
**Update:**
a new Flist is available now with the latest Kubernetes release 1.26 and a few fixes and improvements.
It includes a fix for this issue as we now detect all the public IPs available, and add them as SANs.
https://github.com/threefoldtech/tf-images/pull/122
The update will take place, as soon as the new flist gets promoted to the official apps' repo.
